File: backend-dotnet/FinanceTracker.Api/Services/LedgerService.cs
using FinanceTracker.Api.Data;
using FinanceTracker.Api.Entities;

namespace FinanceTracker.Api.Services;

public sealed class LedgerService(FinanceTrackerDbContext dbContext)
{
    public Transaction CreateTransaction(
        User user,
        Account account,
        Category? category,
        TransactionType type,
        decimal amount,
        DateOnly transactionDate,
        string? merchant,
        string? note,
        string? paymentMethod,
        Guid? transferGroupId)
    {
        ApplyEffect(account, type, amount);

        var transaction = new Transaction
        {
            User = user,
            UserId = user.Id,
            Account = account,
            AccountId = account.Id,
            Category = category,
            CategoryId = category?.Id,
            Type = type,
            Amount = amount,
            TransactionDate = transactionDate,
            Merchant = merchant,
            Note = note,
            PaymentMethod = paymentMethod,
            TransferGroupId = transferGroupId
        };

        dbContext.Transactions.Add(transaction);
        return transaction;
    }

    public void ReverseTransaction(Transaction transaction)
    {
        ReverseEffect(transaction.Account, transaction.Type, transaction.Amount);
    }

    public void ApplyEffect(Account account, TransactionType type, decimal amount)
    {
        account.CurrentBalance = type switch
        {
            TransactionType.INCOME or TransactionType.TRANSFER_IN => account.CurrentBalance + amount,
            TransactionType.EXPENSE or TransactionType.TRANSFER_OUT => account.CurrentBalance - amount,
            _ => account.CurrentBalance
        };
    }

    public void ReverseEffect(Account account, TransactionType type, decimal amount)
    {
        account.CurrentBalance = type switch
        {
            TransactionType.INCOME or TransactionType.TRANSFER_IN => account.CurrentBalance - amount,
            TransactionType.EXPENSE or TransactionType.TRANSFER_OUT => account.CurrentBalance + amount,
            _ => account.CurrentBalance
        };
    }
}

File: backend/src/main/java/com/finance/services/LedgerService.java
package com.finance.services;

import com.finance.entities.Account;
import com.finance.entities.Category;
import com.finance.entities.Transaction;
import com.finance.entities.TransactionType;
import com.finance.entities.User;
import com.finance.repositories.AccountRepository;
import com.finance.repositories.TransactionRepository;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class LedgerService {

    private final AccountRepository accountRepository;
    private final TransactionRepository transactionRepository;

    public Transaction createTransaction(User user,
                                         Account account,
                                         Category category,
                                         TransactionType type,
                                         BigDecimal amount,
                                         LocalDate transactionDate,
                                         String merchant,
                                         String note,
                                         String paymentMethod,
                                         UUID transferGroupId) {
        applyEffect(account, type, amount);
        accountRepository.save(account);

        Transaction transaction = new Transaction();
        transaction.setUser(user);
        transaction.setAccount(account);
        transaction.setCategory(category);
        transaction.setType(type);
        transaction.setAmount(amount);
        transaction.setTransactionDate(transactionDate);
        transaction.setMerchant(merchant);
        transaction.setNote(note);
        transaction.setPaymentMethod(paymentMethod);
        transaction.setTransferGroupId(transferGroupId);
        return transactionRepository.save(transaction);
    }

    public void reverseTransaction(Transaction transaction) {
        Account account = transaction.getAccount();
        reverseEffect(account, transaction.getType(), transaction.getAmount());
        accountRepository.save(account);
    }

    public void applyEffect(Account account, TransactionType type, BigDecimal amount) {
        account.setCurrentBalance(switch (type) {
            case INCOME, TRANSFER_IN -> account.getCurrentBalance().add(amount);
            case EXPENSE, TRANSFER_OUT -> account.getCurrentBalance().subtract(amount);
        });
    }

    public void reverseEffect(Account account, TransactionType type, BigDecimal amount) {
        account.setCurrentBalance(switch (type) {
            case INCOME, TRANSFER_IN -> account.getCurrentBalance().subtract(amount);
            case EXPENSE, TRANSFER_OUT -> account.getCurrentBalance().add(amount);
        });
    }
}
